File: src/main/kotlin/marksync/Marksync.kt
package marksync

import io.github.cdimascio.dotenv.Dotenv
import io.github.cdimascio.dotenv.dotenv
import marksync.remote.RemoteDocument
import marksync.remote.RemoteService
import marksync.remote.esa.EsaService
import marksync.remote.esa.EsaUploader
import marksync.remote.qiita.QiitaService
import marksync.remote.zenn.ZennService
import marksync.uploader.S3Uploader
import marksync.uploader.Uploader
import picocli.CommandLine
import picocli.CommandLine.*
import java.io.File
import java.nio.file.Path
import java.nio.file.Paths
import kotlin.system.exitProcess

@Command(
    name = "marksync",
    subcommands = [HelpCommand::class],
    description = ["Marksync"]
)
class Marksync {
    @Option(names = ["-e", "--env"], paramLabel = "<env>", description = ["Environment name"])
    var envName: String? = null

    @Command(name = "import", description = ["Import documents"])
    @Suppress("unused")
    fun import(
        @Option(names = ["-o", "--output"], required = true, paramLabel = "<dir>", description = ["output directory"])
        output: File
    ): Int {
        runImport(envName, output)
        return 0
    }

    @Command(name = "new", description = ["Create new document metafile"])
    @Suppress("unused")
    fun new(): Int = try {
        runNew(envName)
        0
    } catch (e: MarksyncException) {
        System.err.println(e.message)
        e.code.num
    }

    @Command(name = "status", description = ["Check document update status"])
    @Suppress("unused")
    fun status(
        @Parameters(arity = "0..*", paramLabel = "<target>", description = ["Targets"])
        targets: List<File>?,
    ): Int = try {
        runUpdate(envName, targets ?: listOf(File(".")), checkOnly = true)
        0
    } catch (e: MarksyncException) {
        System.err.println(e.message)
        e.code.num
    }

    @Command(name = "diff", description = ["Differ document update"])
    @Suppress("unused")
    fun diff(
        @Parameters(arity = "0..*", paramLabel = "<target>", description = ["Targets"])
        targets: List<File>?,
    ): Int = try {
        runUpdate(envName, targets ?: listOf(File(".")), checkOnly = true, showDiff = true)
        0
    } catch (e: MarksyncException) {
        System.err.println(e.message)
        e.code.num
    }

    @Command(name = "update", description = ["Update document"])
    @Suppress("unused")
    fun update(
        @Option(names = ["-r", "--recursive"], description = ["Process recursively"])
        recursive: Boolean,

        @Option(names = ["-f", "--force"], description = ["Force update"])
        force: Boolean,

        @Option(names = ["-m", "--message"], paramLabel = "<message>", description = ["Update message"])
        message: String?,

        @Parameters(arity = "0..*", paramLabel = "<target>", description = ["Targets"])
        targets: List<File>?,
    ): Int = try {
        runUpdate(
            envName, targets ?: listOf(File(".")),
            recursive = recursive, force = force, message = message
        )
        0
    } catch (e: MarksyncException) {
        System.err.println(e.message)
        e.code.num
    }

    /**
     * run new sub-command
     */
    private fun runNew(envName: String?) {
        val dotenv = getDotenv(envName)
            ?: throw MarksyncException("fatal: no marksync environment.", MarksyncException.ErrorCode.ENV)
        val service = getService(dotenv)
            ?: throw MarksyncException("fatal: invalid environment.", MarksyncException.ErrorCode.ENV)
        createDocument(File("."), service)
    }

    /**
     * run import sub-command
     */
    private fun runImport(envName: String?, output: File) {
        val dotenv = getDotenv(envName)
            ?: throw MarksyncException("fatal: no marksync environment.", MarksyncException.ErrorCode.ENV)
        val service = getService(dotenv)
            ?: throw MarksyncException("fatal: invalid environment.", MarksyncException.ErrorCode.ENV)
        importAll(output, service)
    }

    /**
     * run status/diff/update sub-command
     */
    private fun runUpdate(
        envName: String?,
        targets: List<File>,
        recursive: Boolean = true,
        force: Boolean = false,
        message: String? = null,
        checkOnly: Boolean = false,
        showDiff: Boolean = false,
    ) {
        val dotenv = getDotenv(envName)
            ?: throw MarksyncException("fatal: no marksync environment.", MarksyncException.ErrorCode.ENV)
        val service = getService(dotenv)
            ?: throw MarksyncException("fatal: invalid environment.", MarksyncException.ErrorCode.ENV)

        // check targets
        targets.forEach { target ->
            if (!target.exists()) {
                throw MarksyncException("target not found: $target", MarksyncException.ErrorCode.TARGET)
            }
        }

        // proceed targets
        val files = targets.flatMap { target ->
            listFiles(target, recursive).filter { it.name == DOCUMENT_FILENAME }.map { it.parentFile }
        }
        if (files.count() >= PREFETCH_THRESHOLD) {
            service.prefetch()
        }
        files.forEach {
            service.sync(it, force, message, checkOnly, showDiff)
        }
    }

    /**
     * Find environment file.
     *
     * @param envName Environment name
     * @return Environment file
     */
    private fun findEnv(envName: String?): File? {
        val envCandidates = mutableListOf<File>()
        var path: Path? = Paths.get(".").toAbsolutePath()
        while (path != null) {
            if (envName != null) {
                envCandidates += path.resolve("$ENV_PREFIX/$envName").toFile()
            } else {
                envCandidates += path.resolve(ENV_PREFIX).toFile()
                envCandidates += path.resolve("$ENV_PREFIX/$ENV_DEFAULT").toFile()
            }
            path = path.parent
        }
        return envCandidates.find { it.exists() && it.isFile }
    }

    /**
     * Get Environment.
     *
     * @param envName Environment name
     * @return Dotenv object
     */
    private fun getDotenv(envName: String?): Dotenv? {
        val env = findEnv(envName) ?: return null
        return dotenv {
            filename = ENV_PREFIX
            directory = "/"
            filename = env.absolutePath
        }
    }

    /**
     * List all files recursively under the directory.
     *
     * @param dir Directory
     * @param recursive Scan files recursively
     * @return Files
     */
    private fun listFiles(dir: File, recursive: Boolean = true): List<File> {
        val these = dir.listFiles()!!.toList()
        return if (!recursive) these else
            these + these.filter { it.isDirectory }.flatMap { listFiles(it) }
    }

    /**
     * Import all documents from service.
     *
     * @param outDir Output directory
     * @param service Service object
     */
    private fun importAll(outDir: File, service: RemoteService) {
        if (outDir.exists()) {
            println("$outDir already exists.")
            return
        }
        service.getDocuments().forEach { (docId: String, doc: RemoteDocument) ->
            println("$docId ${doc.getDocumentUrl()}")
            val dir = File(outDir, docId)
            println("  -> ${dir.absolutePath}")
            dir.mkdirs()
            service.saveMeta(doc, dir)
            doc.saveBody(File(dir, LocalDocument.DOCUMENT_FILENAME))
        }
    }

    private fun createDocument(target: File, service: RemoteService) {
        if (!File(target, DOCUMENT_FILENAME).exists()) {
            println("$DOCUMENT_FILENAME not found.")
            return
        }
        service.createMeta(target)
    }

    /**
     * Get service from env.
     *
     * @param dotenv Environment
     * @return Service
     */
    private fun getService(dotenv: Dotenv): RemoteService? {
        val uploader = getUploader(dotenv)
        val service = dotenv["SERVICE"]
        val serviceName = dotenv["SERVICE_NAME"]
        return when (service) {
            QiitaService.SERVICE_NAME -> {
                val username = dotenv["QIITA_USERNAME"]
                val accessToken = dotenv["QIITA_ACCESS_TOKEN"]
                if (username != null && accessToken != null) {
                    QiitaService(serviceName ?: service, username, accessToken, uploader)
                } else null
            }
            EsaService.SERVICE_NAME -> {
                val team = dotenv["ESA_TEAM"]
                val username = dotenv["ESA_USERNAME"]
                val accessToken = dotenv["ESA_ACCESS_TOKEN"]
                if (team != null && username != null && accessToken != null) {
                    EsaService(serviceName ?: service, team, username, accessToken, uploader)
                } else null
            }
            ZennService.SERVICE_NAME -> {
                val username = dotenv["ZENN_USERNAME"]
                val gitDir = dotenv["ZENN_GIT_DIR"]
                val gitUrl = dotenv["ZENN_GIT_URL"]
                val gitBranch = dotenv["ZENN_GIT_BRANCH"]
                val gitUsername = dotenv["ZENN_GIT_USERNAME"]
                val gitPassword = dotenv["ZENN_GIT_PASSWORD"]
                if (username != null && gitDir != null
                    && gitUrl != null && gitBranch != null && gitUsername != null && gitPassword != null
                ) {
                    ZennService(
                        serviceName ?: service, username,
                        gitDir, gitUrl, gitBranch, gitUsername, gitPassword,
                        uploader
                    )
                } else null
            }
            else -> null
        }
    }

    /**
     * Get uploader from env.
     *
     * @param dotenv Environment
     * @return Uploader
     */
    private fun getUploader(dotenv: Dotenv): Uploader? {
        return if (dotenv["UPLOADER"] == "s3") {
            val bucketName = dotenv["S3_BUCKET_NAME"]
            val prefix = dotenv["S3_PREFIX"]
            val baseUrl = dotenv["S3_BASE_URL"]
            if (bucketName != null && baseUrl != null) {
                S3Uploader(bucketName, prefix, baseUrl, dotenv)
            } else null
        } else if (dotenv["SERVICE"] == "esa") {
            val team = dotenv["ESA_TEAM"]
            val accessToken = dotenv["ESA_ACCESS_TOKEN"]
            if (team != null && accessToken != null) {
                EsaUploader(team, accessToken)
            } else null
        } else {
            null
        }
    }

    companion object {
        const val ENV_PREFIX = ".marksync"
        const val ENV_DEFAULT = "default"
        const val DOCUMENT_FILENAME = "index.md"
        const val PREFETCH_THRESHOLD = 25

        @JvmStatic
        fun run(args: Array<String>) {
            val exitCode = CommandLine(Marksync()).execute(*args)
            exitProcess(exitCode)
        }
    }
}

File: src/marksync/marksync.ts
import {ArgumentParser} from "argparse";
import * as dotenv from "dotenv";
import * as fs from "fs";
import * as path from "path";
import {EsaApiClient} from "./remote/esa/api";
import {EsaUploader} from "./remote/esa/uploader";
import {EsaService} from "./remote/esa/service";
import {QiitaApiClient} from "./remote/qiita/api";
import {QiitaService} from "./remote/qiita/service";
import {ZennRepository} from "./remote/zenn/repository";
import {ZennService} from "./remote/zenn/service";
import {RemoteService} from "./remote/service";
import {S3Uploader} from "./uploader/s3";
import {Uploader} from "./uploader/uploader";
import {MarksyncEnv} from "./env";
import {LocalDocument} from "./document";

const {version} = require("../../package.json");

export class Marksync {
    static readonly ENV_PREFIX = ".marksync";
    static readonly ENV_DEFAULT = "default";
    static readonly PREFETCH_THRESHOLD = 25;

    async run(): Promise<void> {
        const parser = new ArgumentParser({"prog": "marksync"});
        parser.add_argument("-v", "--version", {action: "version", version});
        parser.add_argument("-e", "--env", {help: "environment name"});
        const subparsers = parser.add_subparsers();

        // sub-command: import
        const parser_import = subparsers.add_parser("import", {description: "Import documents"});
        parser_import.add_argument("-o", "--output", {required: true, metavar: "<dir>", help: "output directory"});
        parser_import.set_defaults({
            "func": async () => {
                await this.runImport(args.env, args.output);
            }
        });

        // sub-command: new
        const parser_new = subparsers.add_parser("new", {description: "Create new document metafile"});
        parser_new.set_defaults({
            "func": async () => {
                this.runNew(args.env);
            }
        });

        // sub-command: status
        const parser_status = subparsers.add_parser("status", {description: "Check document update status"});
        parser_status.add_argument("targets", {metavar: "<target>", nargs: "*", default: ["."], help: "targets"});
        parser_status.set_defaults({
            "func": async () => {
                await this.runUpdate(args.env, args.targets, true, false, null, true);
            }
        });

        // sub-command: diff
        const parser_diff = subparsers.add_parser("diff", {description: "Differ document update"});
        parser_diff.add_argument("targets", {metavar: "<target>", nargs: "*", default: ["."], help: "targets"});
        parser_diff.set_defaults({
            "func": async () => {
                await this.runUpdate(args.env, args.targets, true, false, null, true, true);
            }
        });

        // sub-command: update
        const parser_update = subparsers.add_parser("update", {description: "Update document"});
        parser_update.add_argument("-r", "--recursive", {action: "store_true", help: "process recursively"});
        parser_update.add_argument("-f", "--force", {action: "store_true", help: "force update"});
        parser_update.add_argument("-m", "--message", {metavar: "<message>", help: "update message"});
        parser_update.add_argument("targets", {metavar: "<target>", nargs: "*", default: ["."], help: "targets"});
        parser_update.set_defaults({
            "func": async () => {
                await this.runUpdate(args.env, args.targets, args.recursive, args.force, args.message);
            }
        });

        const args = parser.parse_args();
        if (args.func) {
            await args.func(args);
        } else {
            parser.print_usage();
        }
    }

    /**
     * run new sub-command
     */
    runNew(envName: string): void {
        const service = this.getService(this.getDotenv(envName));
        if (!service) return;

        this.createDocument(".", service);
    }

    /**
     * run import sub-command
     */
    async runImport(envName: string, output: string): Promise<void> {
        const service = this.getService(this.getDotenv(envName));
        if (!service) return;

        await this.importAll(output, service);
    }

    /**
     * run status/diff/update sub-command
     */
    async runUpdate(envName: string, targets: string[], recursive: boolean = true, force: boolean = false, message: string | null, checkOnly: boolean = false, showDiff: boolean = false): Promise<void> {
        const service = this.getService(this.getDotenv(envName));
        if (!service) return;

        // check targets
        for (const target of targets) {
            if (!fs.existsSync(target)) {
                process.stderr.write(`target not found: ${target}\n`);
                return;
            }
        }

        // proceed targets
        const files = targets.flatMap(target =>
            this.listFiles(target, recursive)
                .filter(it => path.basename(it) == LocalDocument.DOCUMENT_FILENAME)
                .map(it => path.dirname(it))
                .sort()
        );
        if (files.length >= Marksync.PREFETCH_THRESHOLD) {
            await service.prefetch();
        }
        for (const file of files) {
            await service.sync(file, force, message, checkOnly, showDiff);
        }
    }

    /**
     * Find environment file.
     *
     * @param envName Environment name
     * @return Environment file path
     */
    private findEnv(envName?: string): string | null {
        const envCandidates = [];
        let p = process.cwd();
        while (p.length > 1) {
            if (envName) {
                envCandidates.push(`${p}/${Marksync.ENV_PREFIX}/${envName}`);
            } else {
                envCandidates.push(`${p}/${Marksync.ENV_PREFIX}`);
                envCandidates.push(`${p}/${Marksync.ENV_PREFIX}/${Marksync.ENV_DEFAULT}`);
            }
            p = path.dirname(p);
        }
        return envCandidates.find(it => fs.existsSync(it) && fs.lstatSync(it).isFile()) || null;
    }

    /**
     * Get Environment.
     *
     * @param envName Environment name
     * @return Dotenv object
     */
    private getDotenv(envName?: string): MarksyncEnv {
        const envFilePath = this.findEnv(envName);
        if (!envFilePath) {
            return {} as MarksyncEnv;
        }
        dotenv.config({path: envFilePath});
        return (process.env as unknown) as MarksyncEnv;
    }

    /**
     * List all files recursively under the directory.
     *
     * @param dirPath Directory path
     * @param recursive Scan files recursively
     * @return Files
     */
    private listFiles(dirPath: string, recursive: boolean = true): string[] {
        let result: string[] = [];
        for (const f of fs.readdirSync(dirPath)) {
            const filePath = path.join(dirPath, f);
            if (recursive && fs.statSync(filePath).isDirectory()) {
                result = result.concat(this.listFiles(filePath));
            } else {
                result.push(filePath);
            }
        }
        return result;
    }

    // noinspection JSMethodCanBeStatic
    /**
     * Import all documents from service.
     *
     * @param outDir Output directory
     * @param service Service object
     */
    private async importAll(outDir: string, service: RemoteService): Promise<void> {
        if (fs.existsSync(outDir)) {
            process.stderr.write(`ERROR: ${outDir} already exists.\n`);
            return;
        }

        const documents = await service.getDocuments();
        if (documents === null) {
            return;
        }
        for (const [docId, doc] of Object.entries(documents)) {
            process.stdout.write(`${docId} ${doc.getDocumentUrl()}\n`);
            const dir = path.join(outDir, docId);
            process.stdout.write(`  -> ${dir}\n`);
            fs.mkdirSync(dir, {recursive: true});
            await service.saveMeta(doc, dir);
            doc.saveBody(path.join(dir, LocalDocument.DOCUMENT_FILENAME));
        }
    }

    // noinspection JSMethodCanBeStatic
    private createDocument(target: string, service: RemoteService): void {
        if (!fs.existsSync(`${target}/${LocalDocument.DOCUMENT_FILENAME}`)) {
            process.stderr.write(`ERROR: ${LocalDocument.DOCUMENT_FILENAME} not found.\n`);
            return;
        }
        service.createMeta(target);
    }

    /**
     * Get service from env.
     *
     * @param env Environment
     * @return Service
     */
    private getService(env: MarksyncEnv): RemoteService | null {
        const uploader = this.getUploader(env);
        const service = env.SERVICE;
        const serviceName = env.SERVICE_NAME;
        if (service == QiitaService.SERVICE_NAME) {
            const username = env.QIITA_USERNAME;
            const accessToken = env.QIITA_ACCESS_TOKEN;
            if (username !== undefined && accessToken !== undefined) {
                return new QiitaService(
                    serviceName || service,
                    new QiitaApiClient(username, accessToken),
                    uploader
                );
            }
        } else if (service == EsaService.SERVICE_NAME) {
            const team = env["ESA_TEAM"];
            const username = env.ESA_USERNAME;
            const accessToken = env.ESA_ACCESS_TOKEN;
            if (team !== undefined && username !== undefined && accessToken !== undefined) {
                return new EsaService(
                    serviceName || service,
                    new EsaApiClient(team, username, accessToken),
                    uploader
                );
            }
        } else if (service == ZennService.SERVICE_NAME) {
            const username = env["ZENN_USERNAME"];
            const gitDir = env["ZENN_GIT_DIR"];
            const gitUrl = env["ZENN_GIT_URL"];
            const gitBranch = env["ZENN_GIT_BRANCH"];
            const gitUsername = env["ZENN_GIT_USERNAME"];
            const gitPassword = env["ZENN_GIT_PASSWORD"];
            if (username !== undefined && gitDir !== undefined && gitUrl !== undefined &&
                gitBranch !== undefined && gitUsername !== undefined && gitPassword !== undefined) {
                return new ZennService(
                    serviceName || service,
                    new ZennRepository(username, gitDir, gitUrl, gitBranch, gitUsername, gitPassword),
                    uploader
                );
            }
        }
        process.stderr.write("ERROR: invalid configuration.\n");
        return null
    }

    // noinspection JSMethodCanBeStatic
    /**
     * Get uploader from env.
     *
     * @param env Environment
     * @return Uploader
     */
    private getUploader(env: MarksyncEnv): Uploader | null {
        if (env.UPLOADER == "s3") {
            const bucketName = env.S3_BUCKET_NAME;
            const prefix = env.S3_PREFIX;
            const baseUrl = env.S3_BASE_URL;
            if (bucketName && baseUrl) {
                return new S3Uploader(bucketName, prefix, baseUrl);
            }
        } else if (env.SERVICE == "esa") {
            const team = env.ESA_TEAM!
            const username = env.ESA_USERNAME!
            const accessToken = env.ESA_ACCESS_TOKEN!
            if (team && username && accessToken) {
                return new EsaUploader(new EsaApiClient(team, username, accessToken));
            }
        }
        return null;
    }
}
